README.md
Travis-CI: [![Build Status](https://travis-ci.org/SICKAG/sick_visionary_t.svg?branch=indigo-devel)](https://travis-ci.org/SICKAG/sick_visionary_t)

# sick_visionary_t
Repository for ROS-enabled 3D sensors from Sick.


sick_visionary_t_driver_test/CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package sick_visionary_t_driver
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.0.2 (2016-10-07)
------------------
* rename to visionary_t
* Contributors: Florian Weisshardt

0.0.1 (2016-08-02)
------------------
* fixed warning
* added parameter:prevent_frame_skipping
  f true: prevents skipping of frames and publish everything, otherwise use newest data to publish to ROS world
* using pointer instead of copy constructor (optimization)
* warning instead of debug msg
* fixed receive queue
* added functions to check header + size before parsing
* debug messaging
* check data header
* detach publishing data from network thread
* Renamed launch file to match the driver name
* Updated formatting and adding parameter comment in launch file.
* initial version of sick_visionary_t_driver ROS driver
* Contributors: Florian Weisshardt, Joshua Hampp, Marco Dierschke


sick_visionary_t_driver_test/src/node.cpp
/*!
 *****************************************************************
 * \file
 *
 * \note
 *   Copyright (c) 2015 \n
 *   Fraunhofer Institute for Manufacturing Engineering
 *   and Automation (IPA) \n\n
 *
 * \note
 *   Copyright (c) 2015 \n
 *   SICK AG \n\n
 *
 *****************************************************************
 *
 * \note
 *   ROS package name: sick_visionary_t_driver
 *
 * \author
 *   Author: Joshua Hampp
 *
 * \date Date of creation: 05/21/2015
 *
 *****************************************************************
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 ****************************************************************/
 
#include <sick_visionary_t_driver/driver.h>
#include <ros/ros.h>
#include <image_transport/image_transport.h>
#include <sensor_msgs/image_encodings.h>
#include <sensor_msgs/PointCloud2.h>
#include <sensor_msgs/LaserScan.h>
#include <cv_bridge/cv_bridge.h>
#include <boost/thread.hpp>
#include <sensor_msgs/CameraInfo.h>
#include <std_msgs/ByteMultiArray.h>

/// Flag whether invalid points are to be replaced by NaN
const bool SUPPRESS_INVALID_POINTS = true;

/// Distance code for data outside the TOF range
const uint16_t NARE_DISTANCE_VALUE = 0xffffU;
image_transport::Publisher g_pub_depth, g_pub_confidence, g_pub_intensity;
ros::Publisher g_pub_camera_info, g_pub_points, g_pub_ios, g_pub_distance;
Driver_3DCS::Control *g_control = NULL;
std::string g_frame_id;
/// If true: prevents skipping of frames and publish everything, otherwise use newest data to publish to ROS world
bool g_publish_all = false;

boost::mutex g_mtx_data;
boost::shared_ptr<Driver_3DCS::Data> g_data;

void publish_frame(const Driver_3DCS::Data &data);

void thr_publish_frame() {
	g_mtx_data.lock();
	publish_frame(*g_data);
	g_mtx_data.unlock();
}

void on_frame(const boost::shared_ptr<Driver_3DCS::Data> &data) {
	//update data in queue and
	//detach publishing data from network thread
	
	if(g_publish_all || g_mtx_data.try_lock()) {
		if(g_publish_all)
			g_mtx_data.lock();
		g_data = data;
		g_mtx_data.unlock();
		
		boost::thread thr(thr_publish_frame);
	}
	else
		ROS_WARN("skipping frame");
}

void publish_frame(const Driver_3DCS::Data &data) {
	bool published_anything = false;
	
	sensor_msgs::ImagePtr msg;
	std_msgs::Header header, header_laser;
	header.stamp = ros::Time::now();
	header.frame_id = g_frame_id;
    header_laser.stamp = ros::Time::now();// changed 
    header_laser.frame_id ="camera_rotate";// changed
	
	if(g_pub_camera_info.getNumSubscribers()>0) {
		published_anything = true;
		
		sensor_msgs::CameraInfo ci;
		ci.header = header;
		
		ci.height = data.getCameraParameters().height;
		ci.width  = data.getCameraParameters().width;
		
		ci.D.clear();
		ci.D.resize(5,0);
		ci.D[0] = data.getCameraParameters().k1;
		ci.D[1] = data.getCameraParameters().k2;
		
		for(int i=0; i<9; i++) ci.K[i]=0;
		ci.K[0] = data.getCameraParameters().fx;
		ci.K[4] = data.getCameraParameters().fy;
		ci.K[2] = data.getCameraParameters().cx;
		ci.K[5] = data.getCameraParameters().cy;
		
		for(int i=0; i<12; i++)
			ci.P[i] = 0;//data.getCameraParameters().cam2worldMatrix[i];
		//TODO:....
		ci.P[0] = data.getCameraParameters().fx;
		ci.P[5] = data.getCameraParameters().fy;
		ci.P[10] = 1;
		ci.P[2] = data.getCameraParameters().cx;
		ci.P[6] = data.getCameraParameters().cy;
			
		g_pub_camera_info.publish(ci);
	}
	
	if(g_pub_depth.getNumSubscribers()>0) {
		published_anything = true;
		
		msg = cv_bridge::CvImage(std_msgs::Header(), sensor_msgs::image_encodings::TYPE_16UC1, data.get_distance()).toImageMsg();
		msg->header = header;
		g_pub_depth.publish(msg);
	}
	if(g_pub_confidence.getNumSubscribers()>0) {
		published_anything = true;
		
		msg = cv_bridge::CvImage(std_msgs::Header(), sensor_msgs::image_encodings::TYPE_16UC1, data.get_confidence()).toImageMsg();
		msg->header = header;
		g_pub_confidence.publish(msg);
	}
	if(g_pub_intensity.getNumSubscribers()>0) {
		published_anything = true;
		
		msg = cv_bridge::CvImage(std_msgs::Header(), sensor_msgs::image_encodings::TYPE_16UC1, data.get_intensity()).toImageMsg();
		msg->header = header;
		g_pub_intensity.publish(msg);
	}
	if(g_pub_points.getNumSubscribers()>0) {
		published_anything = true;
		
		typedef sensor_msgs::PointCloud2 PointCloud;
		
		// Allocate new point cloud message
		PointCloud::Ptr cloud_msg (new PointCloud);
		cloud_msg->header = header; // Use depth image time stamp
		cloud_msg->height = data.get_distance().rows;
		cloud_msg->width  = data.get_distance().cols;
		cloud_msg->is_dense = false;
		cloud_msg->is_bigendian = false;

		cloud_msg->fields.resize (5);
		cloud_msg->fields[0].name = "x"; cloud_msg->fields[1].name = "y"; cloud_msg->fields[2].name = "z";
		cloud_msg->fields[3].name = "confidence"; cloud_msg->fields[4].name = "intensity";
		int offset = 0;
		// All offsets are *4, as all field data types are float32
		for (size_t d = 0; d < cloud_msg->fields.size (); ++d, offset += 4)
		{
			cloud_msg->fields[d].offset = offset;
			cloud_msg->fields[d].datatype = (d<3) ? int(sensor_msgs::PointField::FLOAT32) : int(sensor_msgs::PointField::UINT16);
			cloud_msg->fields[d].count  = 1;
		}
		cloud_msg->point_step = offset;
		cloud_msg->row_step   = cloud_msg->point_step * cloud_msg->width;
		cloud_msg->data.resize (cloud_msg->height * cloud_msg->width * cloud_msg->point_step);

        const float f2rc = data.getCameraParameters().f2rc / 1000.f; // since f2rc is given in [mm], but the pcl message wants [m]
		
		uint16_t *pDepth, *pConf, *pInt;
		int cp=0;
		for(int i = 0; i < data.get_distance().rows; ++i)
		{
			pDepth = (uint16_t*)(data.get_distance()  .data + (data.get_distance()  .step*i) );
			pConf  = (uint16_t*)(data.get_confidence().data + (data.get_confidence().step*i) );
			pInt   = (uint16_t*)(data.get_intensity() .data + (data.get_intensity() .step*i) );
			
			for (int j = 0; j < data.get_distance().cols; ++j, ++cp)
			{
				if(pDepth[j]==0 || pDepth[j]==NARE_DISTANCE_VALUE) {
					const float bad_point = std::numeric_limits<float>::quiet_NaN();
					memcpy (&cloud_msg->data[cp * cloud_msg->point_step + cloud_msg->fields[0].offset], &bad_point, sizeof (float));
					memcpy (&cloud_msg->data[cp * cloud_msg->point_step + cloud_msg->fields[1].offset], &bad_point, sizeof (float));
					memcpy (&cloud_msg->data[cp * cloud_msg->point_step + cloud_msg->fields[2].offset], &bad_point, sizeof (float));
				}
				else {
					float xp = (data.getCameraParameters().cx - j) / data.getCameraParameters().fx;
					float yp = (data.getCameraParameters().cy - i) / data.getCameraParameters().fy;

					// Correction of the lens distortion
					float r2 = (xp * xp + yp * yp);
					float r4 = r2 * r2;
					float k = 1 + data.getCameraParameters().k1 * r2 + data.getCameraParameters().k2 * r4;
					xp = xp * k;
					yp = yp * k;

					// Calculation of the distances
					float s0 = std::sqrt(xp*xp + yp*yp + 1.f) * 1000.f;

					// Calculation of the Cartesian coordinates
					const float ox = xp * pDepth[j] / s0;
					const float oy = yp * pDepth[j] / s0;
                    const float oz = pDepth[j] / s0 - f2rc;
					
					memcpy (&cloud_msg->data[cp * cloud_msg->point_step + cloud_msg->fields[0].offset], &ox, sizeof (float));
					memcpy (&cloud_msg->data[cp * cloud_msg->point_step + cloud_msg->fields[1].offset], &oy, sizeof (float));
					memcpy (&cloud_msg->data[cp * cloud_msg->point_step + cloud_msg->fields[2].offset], &oz, sizeof (float));
				}
				
				memcpy (&cloud_msg->data[cp * cloud_msg->point_step + cloud_msg->fields[3].offset], &pConf[j], sizeof (uint16_t));
				memcpy (&cloud_msg->data[cp * cloud_msg->point_step + cloud_msg->fields[4].offset], &pInt [j], sizeof (uint16_t));
			}
		}
		
		g_pub_points.publish(cloud_msg);
	}

	if(g_pub_distance.getNumSubscribers()>0) {
		published_anything = true;
		
		sensor_msgs::LaserScan scan_msg;
		scan_msg.ranges.resize(data.getPolar2DParameters().numPolarValues);
		scan_msg.angle_min = data.getPolar2DParameters().angle_of_first_scan_point*3.14/180.0;//*****************************to check it's pi or degree********************************************//
		scan_msg.angle_max = data.getPolar2DParameters().angle_of_first_scan_point*3.14/180.0 + data.getPolar2DParameters().numPolarValues*data.getPolar2DParameters().angular_resolution*3.14/180.0;
		scan_msg.angle_increment = data.getPolar2DParameters().angular_resolution*3.14/180.0;
		scan_msg.range_min = 0.01;
		scan_msg.range_max = 10.0;//*****************************unit maybe wrong********************************************//
		scan_msg.header = header_laser;//*****************************here must be changed when in lab********************************************//
		std::vector<float> distance_polar2d = data.get_polar2d_distance();
		for(int i = 0; i < data.getPolar2DParameters().numPolarValues; ++i) {
			scan_msg.ranges[i] = distance_polar2d[i]/1000.0;		
		}
		g_pub_distance.publish(scan_msg);
	}
		
	
	if(!published_anything) {
		if(g_control) g_control->stopStream();
	}
}

void _on_new_subscriber() {        
        ROS_DEBUG("Got new subscriber");

	if(!g_control) return;

	bool r=g_control->startStream(); // tell the device to start streaming data
	ROS_ASSERT(r);
}

void on_new_subscriber_ros(const ros::SingleSubscriberPublisher& pub) {        
        _on_new_subscriber();
}

void on_new_subscriber_it(const image_transport::SingleSubscriberPublisher& pub) {        
        _on_new_subscriber();
}

int main(int argc, char **argv) {
	ros::console::set_logger_level(ROSCONSOLE_DEFAULT_NAME, ros::console::levels::Debug);
	ros::init(argc, argv, "driver_3DCS");
	ros::NodeHandle nh("~");
	
	bool r;
	boost::asio::io_service io_service;
	
	//default parameters
	std::string remote_device_ip="192.168.1.10";
	g_frame_id = "camera";
	
	ros::param::get("~remote_device_ip", remote_device_ip);
	ros::param::get("~frame_id", g_frame_id);
	ros::param::get("~prevent_frame_skipping", g_publish_all);
	
	Driver_3DCS::Control control(io_service, remote_device_ip);
	r=control.open();
	ROS_ASSERT(r);
	r=control.initStream();
	ROS_ASSERT(r);
	
	boost::thread thr(boost::bind(&boost::asio::io_service::run, &io_service));
	
	Driver_3DCS::Streaming device(io_service, remote_device_ip);
	device.getSignal().connect( boost::bind(&on_frame, _1) );
	r=device.openStream();
	ROS_ASSERT(r);
	
	g_control = &control;
	
	//make me public (after init.)
	
	image_transport::ImageTransport it(nh);
	g_pub_depth = it.advertise("depth", 1, (image_transport::SubscriberStatusCallback)on_new_subscriber_it, image_transport::SubscriberStatusCallback());
	g_pub_points = nh.advertise<sensor_msgs::PointCloud2>("points", 2, (ros::SubscriberStatusCallback)on_new_subscriber_ros, ros::SubscriberStatusCallback());
	g_pub_confidence = it.advertise("confidence", 1, (image_transport::SubscriberStatusCallback)on_new_subscriber_it, image_transport::SubscriberStatusCallback());
	g_pub_intensity = it.advertise("intensity", 1, (image_transport::SubscriberStatusCallback)on_new_subscriber_it, image_transport::SubscriberStatusCallback());
	g_pub_camera_info = nh.advertise<sensor_msgs::CameraInfo>("camera_info", 1, (ros::SubscriberStatusCallback)on_new_subscriber_ros, ros::SubscriberStatusCallback());
	g_pub_ios = nh.advertise<std_msgs::ByteMultiArray>("ios", 1, (ros::SubscriberStatusCallback)on_new_subscriber_ros, ros::SubscriberStatusCallback());
	g_pub_distance = nh.advertise<sensor_msgs::LaserScan>("laserscan", 1, (ros::SubscriberStatusCallback)on_new_subscriber_ros, ros::SubscriberStatusCallback());
	
	//wait til end of exec.
	ros::spin();

	io_service.stop();
	device.closeStream();
	control.close();
	
	g_pub_depth.shutdown();
	g_pub_confidence.shutdown();
	g_pub_intensity.shutdown();
	g_pub_camera_info.shutdown();
	g_pub_points.shutdown();
	g_pub_ios.shutdown();
	g_pub_distance.shutdown();

	return 0;
}


sick_visionary_t_driver_test/src/driver.cpp
/*!
 *****************************************************************
 * \file
 *
 * \note
 *   Copyright (c) 2015 \n
 *   Fraunhofer Institute for Manufacturing Engineering
 *   and Automation (IPA) \n\n
 *
 * \note
 *   Copyright (c) 2015 \n
 *   SICK AG \n\n
 *
 *****************************************************************
 *
 * \note
 *   ROS package name: sick_visionary_t_driver
 *
 * \author
 *   Author: Joshua Hampp
 *
 * \date Date of creation: 05/21/2015
 *
 *****************************************************************
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 ****************************************************************/
#include <sick_visionary_t_driver/driver.h>


sick_visionary_t_driver_test/include/sick_visionary_t_driver/device.h
/*!
 *****************************************************************
 * \file
 *
 * \note
 *   Copyright (c) 2015 \n
 *   Fraunhofer Institute for Manufacturing Engineering
 *   and Automation (IPA) \n\n
 *
 * \note
 *   Copyright (c) 2015 \n
 *   SICK AG \n\n
 *
 *****************************************************************
 *
 * \note
 *   ROS package name: sick_visionary_t_driver
 *
 * \author
 *   Author: Joshua Hampp
 *
 * \date Date of creation: 05/21/2015
 *
 *****************************************************************
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 ****************************************************************/
 
class CoLaFrame {
	std::vector<char> buffer_;
	
	uint8_t checksum() {
		uint8_t c=0;
		for(size_t i=8; i<buffer_.size(); i++)
			c^=buffer_[i];
		return c;
	}
	
public:
	CoLaFrame() : buffer_(8, (char)0) {
	}
	
	CoLaFrame(const char *str) : buffer_(8+strlen(str), (char)0) {
		int i=8;
		while(*str)
			buffer_[i++] = *(str++);
	}
	
	std::vector<char> finish() {
		char *data = &buffer_[0];
		*(uint32_t*)(data+0) = htonl(0x02020202);
		*(uint32_t*)(data+4) = htonl(buffer_.size()-8);
		
		buffer_.push_back(checksum());
		
		return buffer_;
	}
};


/* all methods that use the control channel (sopas) */
class Control : public TCP_Session {
	Any_Session::SIG_ON_DATA on_data_;		///< signal handler for incoming data
	const std::string remote_device_ip_;
	bool stream_started_;
	
	void on_data(const char *data, const size_t size, Any_Session *writer)
	{
		ROS_DEBUG("got data for control");
	}
public:
	Control(boost::asio::io_service& io_service, const std::string &remote_device_ip) :
		TCP_Session(io_service, on_data_),
		remote_device_ip_(remote_device_ip),
		stream_started_(false)
	{
		on_data_.connect( boost::bind(&Control::on_data, this, _1, _2, _3) );
	}
	
    /* establish the control channel to the device */
    bool open() {
        ROS_DEBUG("Connecting to device...");
        
        if(!connect(remote_device_ip_, "2112")) {
            ROS_ERROR("Error on connecting to %s", remote_device_ip_.c_str());
            return false;
		}
            
        ROS_DEBUG("done.");
        return true;
    }

    /* Tells the device that there is a streaming channel by invoking a
     * method named GetBlobClientConfig.
     */
    bool initStream() {
		
        ROS_DEBUG("Initializing streaming...");
        CoLaFrame fr("sMN GetBlobClientConfig");
        ROS_DEBUG("Sending on sMN GetBlobClientConfig");
        write(fr.finish());
        ROS_DEBUG("done.");
        
        return true;
	}

    /* Start streaming the data by calling the "PLAYSTART" method on the
     * device and sending a "Blob request" afterwards.
     */
    bool startStream() {   
		if(stream_started_) return true;
		     
		CoLaFrame fr("sMN PLAYSTART");
        ROS_DEBUG("Sending on sopas sMN PLAYSTART");
        write(fr.finish());
        stream_started_ = true;
        
        return true;
	}
            
        
    /* Stops the data stream. */
    bool stopStream() { 
        if(!stream_started_) return true;

		CoLaFrame fr("sMN PLAYSTOP");
        ROS_DEBUG("Sending on sopas sMN PLAYSTOP");
        write(fr.finish());
        stream_started_ = false;
        
        return true;
	}
};


    
/* All methods that use the streaming channel. */
class Streaming : public TCP_Session {
public:

	struct SFrame {
		std::vector<char> buffer;
		
		typedef boost::shared_ptr<SFrame> Ptr;
	};
	
	typedef boost::signals2::signal<void (const boost::shared_ptr<Data> &)> SIG_ON_FRAME;
	
private:
	const std::string remote_device_ip_;
	Any_Session::SIG_ON_DATA on_data_;		///< signal handler for incoming data
	SIG_ON_FRAME on_frame_;
	bool debugOutput_;
	SFrame::Ptr cur_frame_;
	
public:

	Streaming(boost::asio::io_service& io_service, const std::string &remote_device_ip) :
		TCP_Session(io_service, on_data_),
		remote_device_ip_(remote_device_ip),
		debugOutput_(false)
	{
		on_data_.connect( boost::bind(&Streaming::on_data, this, _1, _2, _3) );
	}
	
	SIG_ON_FRAME &getSignal() {return on_frame_;}
	
	bool &debugFlag() 		{return debugOutput_;}
	bool debugFlag() const {return debugOutput_;}
	
    /* Opens the streaming channel. */
    bool openStream() {
        ROS_DEBUG("Opening streaming socket...");
        if(!connect(remote_device_ip_, "2114")) {
            ROS_DEBUG("Error on connecting to %s", remote_device_ip_.c_str());
            return false;
		}
        ROS_DEBUG("done.");
        
        // saying hello
        const char HEARTBEAT_MSG[] = "BlbReq";
        
        ROS_DEBUG("Sending BlbReq: %s", HEARTBEAT_MSG);
        write(std::string(HEARTBEAT_MSG, sizeof(HEARTBEAT_MSG)-1));
        
        return true;
	}
        
    /* Closes the streaming channel. */
    bool closeStream() {
        ROS_DEBUG("Closing streaming connection...");
        close();
        cur_frame_.reset();
        ROS_DEBUG("done.");
        
        return true;
	}

    /*
     *  Receives the raw data frame from the device via the streamingbchannel.
     */
	void on_data(const char *data, const size_t size, Any_Session *writer)
	{
        if(debugOutput_)
            ROS_DEBUG("Reading image from stream...");
        
        if(!data || size<0) {
            ROS_DEBUG("Socket not connected, terminating.");
            return;
		}
            
        if(cur_frame_) {
			cur_frame_->buffer.insert(cur_frame_->buffer.end(), data, data+size );

			bool goon=true;
			while(goon) {
				goon=false;

				if(Data::check_header(&cur_frame_->buffer[0], cur_frame_->buffer.size())) {				
					boost::shared_ptr<Data> parser(new Data);
					const size_t actual_size = parser->actual_size(&cur_frame_->buffer[0], cur_frame_->buffer.size());
					if(parser->read(&cur_frame_->buffer[0], actual_size)) {
						on_frame_(parser);
						goon=true;
					}
					else {
						cur_frame_.reset();
						ROS_ERROR("failed to parse frame");
						break;
					}

					cur_frame_->buffer.erase(cur_frame_->buffer.begin(), cur_frame_->buffer.begin()+actual_size);
				}
			}
		}
		else {
			cur_frame_.reset(new SFrame());
			on_data(data, size, writer);
		}
			   
		if(debugOutput_)
			ROS_DEBUG("done.");
	}
};


sick_visionary_t_driver_test/include/sick_visionary_t_driver/driver.h
/*!
 *****************************************************************
 * \file
 *
 * \note
 *   Copyright (c) 2015 \n
 *   Fraunhofer Institute for Manufacturing Engineering
 *   and Automation (IPA) \n\n
 *
 * \note
 *   Copyright (c) 2015 \n
 *   SICK AG \n\n
 *
 *****************************************************************
 *
 * \note
 *   ROS package name: sick_visionary_t_driver
 *
 * \author
 *   Author: Joshua Hampp
 *
 * \date Date of creation: 05/21/2015
 *
 *****************************************************************
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 ****************************************************************/
 
#pragma once
#include "network.h"
#include <opencv2/highgui/highgui.hpp>
#include <boost/property_tree/xml_parser.hpp>
#include <boost/lexical_cast.hpp>
#include <boost/algorithm/string.hpp>
#include <boost/foreach.hpp>
#include <ros/assert.h>

//for 64bit
#if defined(__linux__)
#  include <endian.h>
#elif defined(__FreeBSD__) || defined(__NetBSD__)
#  include <sys/endian.h>
#elif defined(__OpenBSD__)
#  include <sys/types.h>
#  define be16toh(x) betoh16(x)
#  define be32toh(x) betoh32(x)
#  define be64toh(x) betoh64(x)
#endif

namespace Driver_3DCS {
	
#include "data.h"
#include "device.h"

}


sick_visionary_t_driver_test/include/sick_visionary_t_driver/network.h
/*!
 *****************************************************************
 * \file
 *
 * \note
 *   Copyright (c) 2015 \n
 *   Fraunhofer Institute for Manufacturing Engineering
 *   and Automation (IPA) \n\n
 *
 * \note
 *   Copyright (c) 2015 \n
 *   SICK AG \n\n
 *
 *****************************************************************
 *
 * \note
 *   ROS package name: sick_visionary_t_driver
 *
 * \author
 *   Author: Joshua Hampp
 *
 * \date Date of creation: 05/21/2015
 *
 *****************************************************************
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 ****************************************************************/
 
 #pragma once

#include <cstdlib>
#include <iostream>
#include <boost/bind.hpp>
#include <boost/asio.hpp>
#include <boost/signals2/signal.hpp>
#include <boost/thread/mutex.hpp>

#include <ros/console.h>	//for ROS_ERROR, ROS_INFO, ...

//! abstract connection handler to process incoming data (independent of protocol)
class Any_Session
{
public:
  /// incoming data with "data" of size "size" and handler to write back ("writer")
  typedef boost::signals2::signal<void (const char *data, const size_t size, Any_Session *writer)> SIG_ON_DATA;
  
  /// every session needs a data handler
  Any_Session(SIG_ON_DATA &cb)
    : on_data_(cb)
  {
  }
  
  virtual ~Any_Session() {}
	
  /// write a string to the network connection
  virtual void write(const std::string &buffer)=0;
  
protected:

	SIG_ON_DATA &on_data_;	///< signal handler for incoming data
};


//! implementation connection handler for TCP
class TCP_Session : public Any_Session
{
public:
  
  TCP_Session(boost::asio::io_service& io_service, SIG_ON_DATA &cb)
    : Any_Session(cb), socket_(io_service)
  {
  }
  
  virtual ~TCP_Session() {
  }

  /// getter for socket connection
  boost::asio::ip::tcp::socket& socket()
  {
    return socket_;
  }

  virtual bool connect(const std::string &path, const std::string &service)
  {
	  boost::asio::ip::tcp::resolver resolver(socket_.get_io_service());
	  boost::asio::ip::tcp::resolver::query query(path, service);
	  boost::asio::ip::tcp::resolver::iterator endpoint_iterator = resolver.resolve(query);
	  boost::asio::ip::tcp::resolver::iterator end;
	  
	  while (endpoint_iterator != end) {
		  if(connect(*endpoint_iterator))
			return true;
	  }
	  
	  return false;
  }
  
  virtual bool connect(const boost::asio::ip::tcp::resolver::endpoint_type &ep)
  {
    close();
    
    boost::system::error_code error = boost::asio::error::host_not_found;
    socket_.connect(ep, error);
    if (error) {
		ROS_ERROR("connecting failed");
      return false;//throw boost::system::system_error(error);
    }
    
    socket_.async_read_some(boost::asio::buffer(data_, max_length),
        boost::bind(&TCP_Session::handle_read, this,
          boost::asio::placeholders::error,
          boost::asio::placeholders::bytes_transferred));
     return true;
  }
  
  virtual void close() {
    if(socket_.is_open()) socket_.close();
    boost::mutex::scoped_lock lock(mtx_);
  }
  
protected:
	
	/// write async. to TCP socket (implement abstract)
	virtual void write(const std::string &buffer) {
      boost::asio::async_write(socket_,
          boost::asio::buffer( buffer ),
          boost::bind(&TCP_Session::handle_write, this,
            boost::asio::placeholders::error));
	}
	virtual void write(const std::vector<char> &buffer) {
      boost::asio::async_write(socket_,
          boost::asio::buffer( buffer ),
          boost::bind(&TCP_Session::handle_write, this,
            boost::asio::placeholders::error));
	}
	
	/// write async. to TCP socket
	virtual void write(const boost::asio::mutable_buffers_1 &buffer) {
      boost::asio::async_write(socket_,
          buffer,
          boost::bind(&TCP_Session::handle_write, this,
            boost::asio::placeholders::error));
	}

private:

  /// called on incoming data --> decides if an error occured or calls handler
  void handle_read(const boost::system::error_code& error,
      size_t bytes_transferred)
  {
    //boost::mutex::scoped_lock lock(mtx_);
      
      if (!error)
    {
		on_data_(data_, bytes_transferred, this);
    }
    else
    {
		ROS_ERROR("error while reading from socket");
      delete this;
    }
    
    if(socket_.is_open())
		socket_.async_read_some(boost::asio::buffer(data_, max_length),
			boost::bind(&TCP_Session::handle_read, this,
			  boost::asio::placeholders::error,
			  boost::asio::placeholders::bytes_transferred));
  }

  /// called if data is written (error handling)
  void handle_write(const boost::system::error_code& error)
  {
    //boost::mutex::scoped_lock lock(mtx_);
    
    if (error)
    {
		ROS_ERROR("error while writing to socket");
      delete this;
    }
  }

protected:
  boost::asio::ip::tcp::socket socket_;	///< TCP socket
  enum { max_length = 4096 };	///< max length of incoming data (shoul be enough?)
  char data_[max_length];		///< data buffer of max. length
  boost::mutex mtx_;			///< lock operations
};
